---
title: "1_Exploratory_Analysis"
author: "Paula Iller"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

GC is a glucocorticoid receptor (GR), a transcription factor that binds to specific genomic sites to influence gene expression. Traditional views have focused on DNA accessibility as the primary determinant of GR binding. However, recent research highlights the importance of other genomic features, including the co-occupancy of transcription factors (TFs), histone modifications, and the overall chromatin landscape.

Our project aims to investigate the binding dynamics and comprehensive analyses of the GR, JUN, and JUNB transcription factors in the  A549 human epithelial lung cell line following treatment with a GR-activating drug. By leveraging high-quality DNase-seq datasets from Ian C. McDowell et al., we will analyze TF occupancy between time point 1 and time point 12. We focused on how GR interacts with JUN and JUNB, and how these interactions influence genomic binding sites and subsequent gene expression changes.

We will conduct a series of analyses to achieve the following objectives:
1. Examine the binding dynamics of GR, JUN, and JUNB.
2. Identify factors that dictate GR binding specificity in the presence of JUN and JUNB.

Our methods will Data Acquisition, Differential Analysis, Visualization of Differential Peaks, Validation with GR ChIP-seq,Dynamic vs. Non-Dynamic Peaks for each group (GR, JUN, JUNB, DNase) and a pairwise comparison using Jaccard Index in dynamic and non-dynamic peak sets. Also checking the temporal binding dynamics within each ChIP-seq group (GR, JUN and JUNB) to see the Upset plot using the narrowPeak data from time point 1 and 12. 

```{r}
# import useful packages
# Load necessary packages
suppressPackageStartupMessages({
    library(GenomicRanges)
    library(rtracklayer)
    library(ChIPseeker)
    library(TxDb.Hsapiens.UCSC.hg38.knownGene)
    library(BSgenome.Hsapiens.UCSC.hg38)
    library(MotifDb)
    library(TFBSTools)
    library(motifmatchr)
    library(ggplot2)
    library(sechm)
    library(epiwraps)
})

```

```{r}
# Importing data 

base_path <- "/Users/paulailler/Documents/ETH/Master/Bioinformatics/Epigen2024_project/data/peaks/"

# Define the file paths with corresponding variable names

file_paths <- c(
  "GR_t1_rep2.bed.gz",
  "GR_t12_rep2.bed.gz",
  "JUN_t1_rep2.bed.gz",
  "JUN_t12_rep2.bed.gz",
  "JUNB_t1_rep2.bed.gz",
  "JUNB_t12_rep2.bed.gz",
  "DNase_t1_rep2.bed.gz",
  "DNase_t12_rep2.bed.gz"
)

variable_names <- c(
  "peaks_GR_t1",
  "peaks_GR_t12",
  "peaks_JUN_t1",
  "peaks_JUN_t12",
  "peaks_JUNB_t1",
  "peaks_JUNB_t12",
  "peaks_DNase_t1",
  "peaks_DNase_t12"
)


# Import each file individually and store in the list
for (i in 1:length(file_paths)) {
  peaks <- rtracklayer::import(file.path(base_path, file_paths[i]), format = "NarrowPeak")
  seqlevelsStyle(peaks) <- "UCSC" 
  peaks_list[[variable_names[i]]] <- peaks
}

# creation of data frame for t1 and t12 conditions
peak_lengths_df_t1 <- do.call(rbind, list(
  data.frame(Condition = "GR_t1", Length = width(peaks_list[["peaks_GR_t1"]])),
  data.frame(Condition = "JUN_t1", Length = width(peaks_list[["peaks_JUN_t1"]])),
  data.frame(Condition = "JUNB_t1", Length = width(peaks_list[["peaks_JUNB_t1"]])),
  data.frame(Condition = "DNase_t1", Length = width(peaks_list[["peaks_DNase_t1"]]))
))

peak_lengths_df_t12 <- do.call(rbind, list(
  data.frame(Condition = "GR_t12", Length = width(peaks_list[["peaks_GR_t12"]])),
  data.frame(Condition = "JUN_t12", Length = width(peaks_list[["peaks_JUN_t12"]])),
  data.frame(Condition = "JUNB_t12", Length = width(peaks_list[["peaks_JUNB_t12"]])),
  data.frame(Condition = "DNase_t12", Length = width(peaks_list[["peaks_DNase_t12"]]))
))

# Plot histograms of peak lengths for t1
plot_t1 <- ggplot(peak_lengths_df_t1, aes(x = Length, fill = Condition)) +
  geom_histogram(binwidth = 50, alpha = 0.6, position = "identity") +
  theme_minimal() +
  labs(title = "Distribution of Peak Lengths (t1)", x = "Peak Length (bp)", y = "Count") +
  scale_fill_brewer(palette = "Set3") +
  xlim(0, 2000)

# Plot histograms of peak lengths for t12
plot_t12 <- ggplot(peak_lengths_df_t12, aes(x = Length, fill = Condition)) +
  geom_histogram(binwidth = 50, alpha = 0.6, position = "identity") +
  theme_minimal() +
  labs(title = "Distribution of Peak Lengths (t12)", x = "Peak Length (bp)", y = "Count") +
  scale_fill_brewer(palette = "Set3") +
  xlim(0, 2000)


# Load the hg38 reference genome
genome <- BSgenome.Hsapiens.UCSC.hg38

# Motif analysis for JUN
JUNmotifs <- MotifDb::query(MotifDb, c("JUN"), "Hsapiens")
JUNmotif <- JUNmotifs[["Hsapiens-HOCOMOCOv10-JUN_HUMAN.H10MO.A"]]
view_motifs(JUNmotif)
JUNmotif2 <- convert_motifs(JUNmotif, class="TFBSTools-PWMatrix")

# Match motifs in DNase-seq peaks
JUN_moi_t1 <- matchMotifs(JUNmotif2, subject = peaks_list[["peaks_DNase_t1"]], genome = genome)
JUN_moi_t1 <- as(setNames(rowRanges(JUN_moi_t1), names(peaks_list[["peaks_DNase_t1"]])), "GRanges")

JUN_moi_t12 <- matchMotifs(JUNmotif2, subject = peaks_list[["peaks_DNase_t12"]], genome = genome)
JUN_moi_t12 <- as(setNames(rowRanges(JUN_moi_t12), names(peaks_list[["peaks_DNase_t12"]])), "GRanges")


# Motif analysis for JUN
JUNBmotifs <- MotifDb::query(MotifDb, c("JUNB"), "Hsapiens")
JUNBmotif <- JUNBmotifs[["Hsapiens-HOCOMOCOv11-core-A-JUNB_HUMAN.H11MO.0.A"]]
view_motifs(JUNBmotif)

JUNBmotif2 <- convert_motifs(JUNBmotif, class="TFBSTools-PWMatrix")

# Match motifs
JUNB_moi_t1 <- matchMotifs(JUNBmotif2, subject = peaks_list[["peaks_DNase_t1"]], genome = genome)
JUNB_moi_t1 <- as(setNames(rowRanges(JUNB_moi_t1), names(peaks_list[["peaks_DNase_t1"]])), "GRanges")

# Convert to GRanges
JUNB_moi_t12 <- matchMotifs(JUNBmotif2, subject = peaks_list[["peaks_DNase_t12"]], genome = genome)
JUNB_moi_t12 <- as(setNames(rowRanges(JUNB_moi_t12), names(peaks_list[["peaks_DNase_t12"]])), "GRanges")



gr_bw1 <- "../data/aligned/GR_t1_rep1.bigWig"  
gr_bed1 <- "../data/peaks/GR_t1_rep2.bed"
gr_bw12 <- "../data/aligned/GR_t12_rep1.bigWig"  
gr_bed12 <- "../data/peaks/GR_t12_rep2.bed"
jun_bw1 <- "../data/aligned/JUN_t1_rep1.bigWig"  
jun_bed1 <- "../data/peaks/JUN_t1_rep2.bed"
jun_bw12 <- "../data/aligned/JUN_t12_rep1.bigWig"  
jun_bed12 <- "../data/peaks/JUN_t12_rep2.bed"

junb_bw1 <- "../data/aligned/JUNB_t1_rep1.bigWig"
junb_bed1 <- "../data/peaks/JUNB_t1_rep2.bed"
junb_bw12 <- "../data/aligned/JUNB_t12_rep1.bigWig"
junb_bed12 <- "../data/peaks/JUNB_t12_rep2.bed"

# Import regions from BED files
gr_regions1 <- rtracklayer::import(gr_bed1,format = "BED")
gr_regions12 <- rtracklayer::import(gr_bed12)
jun_regions1 <- rtracklayer::import(jun_bed1)
jun_regions12 <- rtracklayer::import(jun_bed12)
junb_regions1 <- rtracklayer::import(junb_bed1)
junb_regions12 <- rtracklayer::import(junb_bed12)

# Obtain the matrix of the signal around the regions
gr_matrix1 <- signal2Matrix(gr_bw1, gr_regions1)
gr_matrix12 <- signal2Matrix(gr_bw12, gr_regions12)
jun_matrix1 <- signal2Matrix(jun_bw1, jun_regions1)
jun_matrix12 <- signal2Matrix(jun_bw12, jun_regions12)
junb_matrix1 <- signal2Matrix(junb_bw1, junb_regions1)
junb_matrix12 <- signal2Matrix(junb_bw12, junb_regions12)


# Plot heatmaps
plotEnrichedHeatmaps(gr_matrix1, main = "GR t1")
plotEnrichedHeatmaps(gr_matrix12, main = "GR t12")
plotEnrichedHeatmaps(jun_matrix1, main = "JUN t1")
plotEnrichedHeatmaps(jun_matrix12, main = "JUN t12")
plotEnrichedHeatmaps(junb_matrix1, main = "JUNB t1")
plotEnrichedHeatmaps(junb_matrix12, main = "JUNB t12")
