---
title: "1_Exploratory_Analysis"
author: "Paula Iller"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

GC is a glucocorticoid receptor (GR), a transcription factor that binds to specific genomic sites to influence gene expression. Traditional views have focused on DNA accessibility as the primary determinant of GR binding. However, recent research highlights the importance of other genomic features, including the co-occupancy of transcription factors (TFs), histone modifications, and the overall chromatin landscape.

Our project aims to investigate the binding dynamics of the GR, JUN, and JUNB transcription factors in the A549 human epithelial lung cell line following treatment with a GR-activating drug. By leveraging high-quality ChIP-seq and DNase-seq datasets from Ian C. McDowell et al., we will analyze TF occupancy and chromatin accessibility over a 12-hour period of GC exposure. We focused on how GR interacts with JUN and JUNB, and how these interactions influence genomic binding sites and subsequent gene expression changes.

We will conduct a series of analyses to achieve the following objectives:
1. Examine the binding dynamics of GR, JUN, and JUNB.
2. Identify factors that dictate GR binding specificity in the presence of JUN and JUNB.
3. Correlate TF binding changes with alterations in gene expression and chromatin state.

Our methods will include peak calling and annotation; temporal and statistical analyses of TF binding and chromatin accessibility; co-occurrence analyses; and correlation studies using heatmaps and clustering. We will also perform longitudinal analyses of GR binding dynamics, overlap quantification of early and late GR binding sites, and differential accessibility profiling.


```{r}
# import useful packages
# Load necessary packages
suppressPackageStartupMessages({
    library(motifmatchr)
    library(Biostrings)
    library(MotifDb) 
    library(BSgenome.Hsapiens.UCSC.hg38)
    library(memes)
    library(data.table)
    library(universalmotif)
    library(GenomicRanges)
    library(AnnotationHub)
    library(ensembldb)
    library(InteractionSet)
    library(epiwraps)
    library(ggplot2)
    library(PWMEnrich)
    library(rtracklayer)
    library(ChIPseeker)
    library(org.Hs.eg.db)
    library(TxDb.Hsapiens.UCSC.hg38.knownGene)
    library(ChIPpeakAnno)
    library(VennDiagram)
    library(DiffBind)
    library(motifTestR)
    library(extraChIPs)
    library(rtracklayer)
    library(BSgenome.Hsapiens.UCSC.hg19)
    library(parallel)
    library(ggplot2)
    library(patchwork)
    library(motifmatchr)
    library(Biostrings) # for handling sequences
    library(MotifDb) 
    library(BSgenome.Mmusculus.UCSC.mm10)
    library(memes)
    library(data.table)
    library(universalmotif)
})

```

```{r}
# Importing data 

base_path <- "/Users/paulailler/Documents/ETH/Master/Bioinformatics/Epigen2024_project/data/peaks/"
# Define the file paths with corresponding variable names
file_paths <- c(
  "GR_t1_rep2.bed.gz",
  "GR_t12_rep2.bed.gz",
  "JUN_t1_rep2.bed.gz",
  "JUN_t12_rep2.bed.gz",
  "JUNB_t1_rep2.bed.gz",
  "JUNB_t12_rep2.bed.gz",
  "DNase_t1_rep2.bed.gz",
  "DNase_t12_rep2.bed.gz"
)

variable_names <- c(
  "peaks_GR_t1",
  "peaks_GR_t12",
  "peaks_JUN_t1",
  "peaks_JUN_t12",
  "peaks_JUNB_t1",
  "peaks_JUNB_t12",
  "peaks_DNase_t1",
  "peaks_DNase_t12"
)


# Import each file individually and store in the list
for (i in 1:length(file_paths)) {
  peaks <- rtracklayer::import(file.path(base_path, file_paths[i]), format = "NarrowPeak")
  seqlevelsStyle(peaks) <- "UCSC" 
  peaks_list[[variable_names[i]]] <- peaks
}

# creation of data frame for t1 and t12 conditions
peak_lengths_df_t1 <- do.call(rbind, list(
  data.frame(Condition = "GR_t1", Length = width(peaks_list[["peaks_GR_t1"]])),
  data.frame(Condition = "JUN_t1", Length = width(peaks_list[["peaks_JUN_t1"]])),
  data.frame(Condition = "JUNB_t1", Length = width(peaks_list[["peaks_JUNB_t1"]])),
  data.frame(Condition = "DNase_t1", Length = width(peaks_list[["peaks_DNase_t1"]]))
))

peak_lengths_df_t12 <- do.call(rbind, list(
  data.frame(Condition = "GR_t12", Length = width(peaks_list[["peaks_GR_t12"]])),
  data.frame(Condition = "JUN_t12", Length = width(peaks_list[["peaks_JUN_t12"]])),
  data.frame(Condition = "JUNB_t12", Length = width(peaks_list[["peaks_JUNB_t12"]])),
  data.frame(Condition = "DNase_t12", Length = width(peaks_list[["peaks_DNase_t12"]]))
))

# Plot histograms of peak lengths for t1
plot_t1 <- ggplot(peak_lengths_df_t1, aes(x = Length, fill = Condition)) +
  geom_histogram(binwidth = 50, alpha = 0.6, position = "identity") +
  theme_minimal() +
  labs(title = "Distribution of Peak Lengths (t1)", x = "Peak Length (bp)", y = "Count") +
  scale_fill_brewer(palette = "Set3") +
  xlim(0, 2000)

# Plot histograms of peak lengths for t12
plot_t12 <- ggplot(peak_lengths_df_t12, aes(x = Length, fill = Condition)) +
  geom_histogram(binwidth = 50, alpha = 0.6, position = "identity") +
  theme_minimal() +
  labs(title = "Distribution of Peak Lengths (t12)", x = "Peak Length (bp)", y = "Count") +
  scale_fill_brewer(palette = "Set3") +
  xlim(0, 2000)

# Load the hg38 reference genome
genome <- BSgenome.Hsapiens.UCSC.hg38

# Motif analysis for JUN
JUNmotifs <- MotifDb::query(MotifDb, c("JUN"), "Hsapiens")
JUNmotif <- JUNmotifs[["Hsapiens-HOCOMOCOv10-JUN_HUMAN.H10MO.A"]]
view_motifs(JUNmotif)

JUNmotif2 <- convert_motifs(JUNmotif, class="TFBSTools-PWMatrix")
moi <- motifmatchr::matchMotifs(JUNmotif2, chr1, out="positions", p.cutoff=1e-5)
moi <- as(setNames(moi,names(chr19)), "GRanges")

# Match motifs in DNase-seq peaks
JUN_moi_t1 <- matchMotifs(JUNmotif2, subject = peaks_list[["peaks_DNase_t1"]], genome = genome)
JUN_moi_t12 <- matchMotifs(JUNmotif2, subject = peaks_list[["peaks_DNase_t12"]], genome = genome)

# Motif analysis for JUN
JUNBmotifs <- MotifDb::query(MotifDb, c("JUNB"), "Hsapiens")
JUNBmotif <- JUNBmotifs[["Hsapiens-HOCOMOCOv11-core-A-JUNB_HUMAN.H11MO.0.A"]]
view_motifs(JUNBmotif)

JUNBmotif2 <- convert_motifs(JUNBmotif, class="TFBSTools-PWMatrix")

JUNB_moi_t1 <- matchMotifs(JUNBmotif2, subject = peaks_list[["peaks_DNase_t1"]], genome = genome)
JUNB_moi_t1 <- matchMotifs(JUNBmotif2, subject = peaks_list[["peaks_DNase_t12"]], genome = genome)



# 
# annotated_GR_t1 <- epiwraps::annotateRegions(peaks_list[["peaks_GR_1"]], genome)
# annotated_GR_t12 <- epiwraps::annotateRegions(peaks_list[["peaks_GR_t12"]], genome)
# annotated_JUN_t1 <- epiwraps::annotateRegions(peaks_list[["peaks_JUN_t1"]], genome)
# annotated_JUN_t12 <- epiwraps::annotateRegions(peaks_list[["peaks_JUN_t12"]], genome)
# annotated_JUNB_t1 <- epiwraps::annotateRegions(peaks_list[["peaks_JUNB_t1"]], genome)
# annotated_JUNB_t12 <- epiwraps::annotateRegions(peaks_list[["peaks_JUNB_t12"]], genome)
# 
# # Display annotated peaks
# head(annotated_GR_t1)
# head(annotated_GR_t12)
# head(annotated_JUN_t1)
# head(annotated_JUN_t12)
# head(annotated_JUNB_t1)
# head(annotated_JUNB_t12)
