---
title: "4_Temporal_Binding_Dynamics"
author: "Zhihan Zhu"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup4 setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# Temporal Binding Dynamics

In the pairwise comparisons analysis, we found that most of the colocated binding sites are found in non-dynamic peaks. We inferred that, this is because most of the TF dynamics happen between time point 0 and 1, and thus these differential binding sites can not be significantly detected if we do the differential analysis between time point 1 and 12, and actually aggregated in non-significant group. In this chapter, we are going to compare the peaks at different time points to verify this statement.

```{r}
# import useful packages
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(epiwraps)
  library(GenomicRanges)
  library(rtracklayer)
  library(Rsubread)
  library(R.utils)
  library(ggplot2)
})
```

```{r}
# import peaks, only use replicate 2 for consistency and simplicity

# GR
GR_t1 <- rtracklayer::import("data/peaks/GR_t1_rep2.bed.gz", format = "narrowPeak")
GR_t12 <- rtracklayer::import("data/peaks/GR_t12_rep2.bed.gz", format = "narrowPeak")

# JUN
JUN_t1 <- rtracklayer::import("data/peaks/JUN_t1_rep2.bed.gz", format = "narrowPeak")
JUN_t12 <- rtracklayer::import("data/peaks/JUN_t12_rep2.bed.gz", format = "narrowPeak")

# JUNB
JUNB_t1 <- rtracklayer::import("data/peaks/JUNB_t1_rep2.bed.gz", format = "narrowPeak")
JUNB_t12 <- rtracklayer::import("data/peaks/JUNB_t12_rep2.bed.gz", format = "narrowPeak")

# DNase
DNase_t1 <- rtracklayer::import("data/peaks/DNase_t1_rep2.bed.gz", format = "narrowPeak")
DNase_t12 <- rtracklayer::import("data/peaks/DNase_t12_rep2.bed.gz", format = "narrowPeak")
```

### Use Upset plot to see the intersection

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Upset Plot for Intersect between Different Time Points of GR')}
# GR
suppressWarnings({
  GR_peaks <- list(GR_t1, GR_t12)
  names(GR_peaks) <- c("GR_t1", "GR_t12")
  regionUpset(GR_peaks)
})
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Upset Plot for Intersect between Different Time Points of JUN')}
# JUN
suppressWarnings({
  JUN_peaks <- list(JUN_t1, JUN_t12)
  names(JUN_peaks) <- c("JUN_t1", "JUN_t12")
  regionUpset(JUN_peaks)
})
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Upset Plot for Intersect between Different Time Points of JUNB')}
# JUNB
suppressWarnings({
  JUNB_peaks <- list(JUNB_t1, JUNB_t12)
  names(JUNB_peaks) <- c("JUNB_t1", "JUNB_t12")
  regionUpset(JUNB_peaks)
})
```

It is true that most of the GR binding sites present at later time point (i.e. time point 12) were also present at early time point (i.e. time point 1). However, this is not true though, for JUN and JUNB. It is obvious that, for JUN and JUNB, the binding sites present at later time point are much more than that at early time point, and opposite to GR, nearly all of the JUN and JUNB binding sites present at early time point were also present at later time point.

So it's only true that GR dynamics happen between time point 0 and 1, but not the case for JUN and JUNB.

## Co-occurrence analysis

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Upset Plot for Intersect between Different ChIP-seq peaks at t1')}
# t1
suppressWarnings({
  t1_peaks <- list(GR_t1, JUN_t1, JUNB_t1)
  names(t1_peaks) <- c("GR_t1", "JUN_t1", "JUNB_t1")
  regionUpset(t1_peaks, set_size.show = TRUE)
})
```

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Upset Plot for Intersect between Different ChIP-seq peaks at t12')}
# t12
suppressWarnings({
  t12_peaks <- list(GR_t12, JUN_t12, JUNB_t12)
  names(t12_peaks) <- c("GR_t12", "JUN_t12", "JUNB_t12")
  regionUpset(t12_peaks, set_size.show = TRUE)
})
```

At the beginning of the treatment (time point 1), the co-occurrence of GR, JUN and JUNB happens quite often (4272 out of 8874 GR peaks). At late stage of the treatment (time point 12), the absolute number of co-occurrence peaks is decreasing, however it becomes the most part of GR peaks (3787 out of 3930 peaks). The co-occurrence of JUN and JUNB happens a lot at time point 12.

We are interested in, to what extent JUN/JUNB colocalize within the GR sites.

```{r}
# t1
JUN_GR_t1 <- intersect(JUN_t1, GR_t1)
JUNB_GR_t1 <- intersect(JUNB_t1, GR_t1)

print("At time point 1:")

JUN_t1_prop <- round(sum(overlapsAny(JUN_GR_t1, JUNB_GR_t1))/length(JUN_GR_t1), digits = 4)*100
print(paste0("Of the JUN sites within GR sites, ", as.character(JUN_t1_prop),
             "% overlap JUNB sites within GR sites."))

JUNB_t1_prop <- round(sum(overlapsAny(JUNB_GR_t1, JUN_GR_t1))/length(JUNB_GR_t1), digits = 4)*100
print(paste0("Of the JUNB sites within GR sites, ", as.character(JUNB_t1_prop),
             "% overlap JUN sites within GR sites."))
```

```{r}
# t12
JUN_GR_t12 <- intersect(JUN_t12, GR_t12)
JUNB_GR_t12 <- intersect(JUNB_t12, GR_t12)

print("At time point 12:")

suppressWarnings({
  JUN_t12_prop <- round(sum(overlapsAny(JUN_GR_t12, JUNB_GR_t12))/length(JUN_GR_t12), digits = 4)*100
})
print(paste0("Of the JUN sites within GR sites, ", as.character(JUN_t12_prop),
             "% overlap JUNB sites within GR sites."))

suppressWarnings({
  JUNB_t12_prop <- round(sum(overlapsAny(JUNB_GR_t12, JUN_GR_t12))/length(JUNB_GR_t12), digits = 4)*100
})
print(paste0("Of the JUNB sites within GR sites, ", as.character(JUNB_t12_prop),
             "% overlap JUN sites within GR sites."))
```

### Differential accessibility between different groups of GR sites

Here we would like to separate GR sites into two groups: the ones that gain JUN/JUNB sites, and the ones do not. And we are interested in the GR sites that gain JUN/JUNB show more change in accessibility than that don't.

```{r}
# import GR peak sets at t1
GR_t1_peaks <- c("GR_t1_rep1.bed.gz", "GR_t1_rep2.bed.gz", "GR_t1_rep3.bed.gz")
GR_t1_peaks <- lapply(file.path("data/peaks", GR_t1_peaks),
                      rtracklayer::import, format="narrowPeak")
GR_t1_peaks <- reduce(unlist(GRangesList(GR_t1_peaks)), with.revmap=TRUE)
# length(GR_t1_peaks) 11740

# import JUN/JUNB peak sets at t1
JUN_JUNB_t1_peaks <- c("JUN_t1_rep2.bed.gz", "JUN_t1_rep3.bed.gz",
                       "JUNB_t1_rep2.bed.gz", "JUNB_t1_rep3.bed.gz")
JUN_JUNB_t1_peaks <- lapply(file.path("data/peaks", JUN_JUNB_t1_peaks),
                            rtracklayer::import, format="narrowPeak")
JUN_JUNB_t1_peaks <- reduce(unlist(GRangesList(JUN_JUNB_t1_peaks)), with.revmap=TRUE)
# length(JUN_JUNB_t1_peaks) 35808

# subset GR peaks that do not have JUN/JUNB at t1
suppressWarnings({
  GR_t1_peaks <- subsetByOverlaps(GR_t1_peaks, JUN_JUNB_t1_peaks, invert = FALSE)
})
# length(GR_t1_peaks) 9495

# import GR peak sets at t12
GR_t12_peaks <- c("GR_t12_rep1.bed.gz", "GR_t12_rep2.bed.gz", "GR_t12_rep3.bed.gz")
GR_t12_peaks <- lapply(file.path("data/peaks", GR_t12_peaks),
                       rtracklayer::import, format="narrowPeak")
GR_t12_peaks <- reduce(unlist(GRangesList(GR_t12_peaks)), with.revmap=TRUE)
# length(GR_t12_peaks) 5591

# import JUN/JUNB peak sets at t12
JUN_JUNB_t12_peaks <- c("JUN_t12_rep1.bed.gz", "JUN_t12_rep2.bed.gz",
                        "JUN_t12_rep3.bed.gz", "JUNB_t12_rep1.bed.gz", "JUNB_t12_rep2.bed.gz")
JUN_JUNB_t12_peaks <- lapply(file.path("data/peaks", JUN_JUNB_t12_peaks),
                             rtracklayer::import, format="narrowPeak")
JUN_JUNB_t12_peaks <- reduce(unlist(GRangesList(JUN_JUNB_t12_peaks)), with.revmap=TRUE)
# length(JUN_JUNB_t12_peaks) 67785

# subset GR peaks that have JUN/JUNB at t12
GR_t12_peaks_group1 <- subsetByOverlaps(GR_t12_peaks, JUN_JUNB_t12_peaks)
# length(GR_t12_peaks_group1) 5521
# subset GR peaks that do not have JUN/JUNB at t12
GR_t12_peaks_group2 <- subsetByOverlaps(GR_t12_peaks, JUN_JUNB_t12_peaks, invert = TRUE)
# length(GR_t12_peaks_group2) 70

# GR sites that gain JUN/JUNB at t12 (group 1)
GR_group1 <- subsetByOverlaps(GR_t1_peaks, GR_t12_peaks_group1)
# length(GR_group1) 4761
# GR sites that do not gain JUN/JUNB at t12 (group 2)
GR_group2 <- subsetByOverlaps(GR_t1_peaks, GR_t12_peaks_group2)
# length(GR_group2) 19

# import DNase peaks at t1
DNase_t1_peaks <- c("DNase_t1_rep1.bed.gz", "DNase_t1_rep2.bed.gz",
                    "DNase_t1_rep3.bed.gz", "DNase_t1_rep4.bed.gz", "DNase_t1_rep5.bed.gz")
DNase_t1_peaks <- lapply(file.path("data/peaks", DNase_t1_peaks),
                         rtracklayer::import, format="narrowPeak")
DNase_t1_peaks <- reduce(unlist(GRangesList(DNase_t1_peaks)), with.revmap=TRUE)

# import DNase peaks at t12
DNase_t12_peaks <- c("DNase_t12_rep1.bed.gz", "DNase_t12_rep2.bed.gz",
                     "DNase_t12_rep3.bed.gz", "DNase_t12_rep5.bed.gz")
DNase_t12_peaks <- lapply(file.path("data/peaks", DNase_t12_peaks),
                          rtracklayer::import, format="narrowPeak")
DNase_t12_peaks <- reduce(unlist(GRangesList(DNase_t12_peaks)), with.revmap=TRUE)

# subset GR sites that have accessibility at t1
GR_group1_t1 <- subsetByOverlaps(GR_group1, DNase_t1_peaks)
# length(GR_group1_t1) 4379
GR_group2_t1 <- subsetByOverlaps(GR_group2, DNase_t1_peaks)
# length(GR_group2_t1) 8

# subset GR sites that have accessibility at t12
GR_group1_t12 <- subsetByOverlaps(GR_group1, DNase_t12_peaks)
# length(GR_group1_t12) 4412
GR_group2_t12 <- subsetByOverlaps(GR_group2, DNase_t12_peaks)
# length(GR_group2_t12) 7
```

Basically, if we focus on two groups of GR sites, (1) the ones that do not have JUN/JUNB sites at t1 but gain JUN/JUNB at t12; (2) the ones that do not have JUN/JUNB sites at t1 and do not gain JUN/JUNB at t12, we can see that group 1 has more accessibility at both t1 and t12 than group 2. However, it seems like between t1 and t12, the accessibility of both groups do not change much.

However, the number of group 2 is too small to make a conclusion. Most of the GR sites have JUN/JUNB sites at t12 (5521 out of 5591), and only 70 sites do not have JUN/JUNB sites at t12.

```{r, out.width=c('80%'), fig.align='center', fig.cap=c('Numbers of accessible sites for different GR sites at different time points')}
# create data frame to make barplot
data <- data.frame(
  Group = factor(c("Group 1", "Group 1", "Group 2", "Group 2"),
                 levels = c("Group 1", "Group 2")),
  Time = factor(c("Time 1", "Time 12", "Time 1", "Time 12"),
                levels = c("Time 1", "Time 12")),
  Values = c(length(GR_group1_t1), length(GR_group1_t12),
             length(GR_group2_t1), length(GR_group2_t12))
)

# stacked bar plot
plot <- ggplot(data, aes(x = Group, y = log(Values), fill = Time)) +
  geom_bar(stat = "identity") +
  labs(title = "Numbers of accessible sites for different GR sites at different time points") +
  theme_minimal()
plot
```

Here is a stacked bar plot showing the number of accessible sites for different GR sites at different time points, with the groups stating above. The y axis is in log scale for better visualization.

```{r}
# simple statistical analysis, McNemar test, group 1

# number of sites that are both accessible at t1 and t12
a = sum(overlapsAny(GR_group1_t1, GR_group1_t12)) #4272
# number of sites that are accessible at t1 but not at t12
b = length(GR_group1_t1) - a #107
# number of sites that are not accessible at t1 but accessible at t12
c = length(GR_group1_t12) - a #140
# number of sites that are not accessible at both t1 and t12
d = length(GR_group1) - a - b - c #242

contingency_table = matrix(c(a, b, c, d), nrow = 2, byrow = TRUE)
mcnemar.test(contingency_table)
```

```{r}
# simple statistical analysis, McNemar test, group 2

# number of sites that are both accessible at t1 and t12
a = sum(overlapsAny(GR_group2_t1, GR_group2_t12)) #6
# number of sites that are accessible at t1 but not at t12
b = length(GR_group2_t1) - a #2
# number of sites that are not accessible at t1 but accessible at t12
c = length(GR_group2_t12) - a #1
# number of sites that are not accessible at both t1 and t12
d = length(GR_group2) - a - b - c #10

contingency_table = matrix(c(a, b, c, d), nrow = 2, byrow = TRUE)
mcnemar.test(contingency_table)

```

Actually the sample size for group 2 is too small that we can hardly get any statistical results from it. For group 1 we can see somewhat significant results (p<0.05), but I think it's not strong enough to make a conclusion. To summarize, in both groups, the changes of accessibility between t1 and t12 are not obvious. This is somewhat consistent with the statement in the original paper that "chromatin accessibility does not predetermine GR binding".


# Conclusion and Discussion

In our project, we are interested in the dynamics of GR, JUN and JUNB ChIP-seq and also of the chromatin accessibility during the treatment of dexamethasone, which is a GR-activating drug.

We investigated the differential expression of GR, JUN, JUNB and DNase-seq data between time point 12 and time point 1. We found that most of the differential GR sites are actually decreased. We then tested the differential expression of GR ChIP-seq data between time point 12 and time point 0, and found that most of the sites are increased. This result demonstrates that the TF activity of GR is actually activated at the very beginning of synthetic GC dexamethasone exposure, i.e. between the time point 0 and time point 1, which is consistent with the original paper.

We calculated the Jaccard Index for the dynamic and non-dynamic peaks, and visualized the results using heatmaps. We did not find a significant difference between the dynamic and non-dynamic peaks, which may be due to the fact that we used a different method to identify the dynamic peaks, or that the data we used was different from the original paper.

We compared the binding sites of GR, JUN and JUNB at different time points. We found that, for GR, most of the binding sites at later time point were also present at early time point, and most of the dynamics happen between time point 0 and 1. However, this is not the case for JUN and JUNB. We also found that, the co-occurrence of GR, JUN and JUNB happens quite often at the beginning of the treatment, and the absolute number of co-occurrence peaks is decreasing at late stage of the treatment, however it becomes the most part of GR peaks.

We are also interested in, to what extent JUN/JUNB colocalize within the GR sites, and found that the overlap is quite high. We also separated GR sites into two groups: the ones that gain JUN/JUNB sites, and the ones do not. We found that the group that gain JUN/JUNB sites show more accessibility at both time points than the group that do not. However, the changes of accessibility between time points are not obvious. We also performed statistical analysis to verify this statement, and found somewhat significant results for group 1, but not for group 2. To what extent the accessibility of GR sites is determined by the presence of JUN/JUNB sites is still an open question.

# Reference

McDowell I C, Barrera A, D'Ippolito A M, et al. Glucocorticoid receptor recruits to enhancers and drives activation by motif-directed binding[J]. Genome research, 2018, 28(9): 1272-1284.

Biddie S C, John S, Sabo P J, et al. Transcription factor AP1 potentiates chromatin accessibility and glucocorticoid receptor binding[J]. Molecular cell, 2011, 43(1): 145-155.

John S, Sabo P J, Thurman R E, et al. Chromatin accessibility pre-determines glucocorticoid receptor binding patterns[J]. Nature genetics, 2011, 43(3): 264-268.
